### ISSUE TITLE
Compile OS X 10.9 MATLAB 2014a

### ISSUE THREAD

**Issue by davesc** [kyamagu/mexopencv#110]
I get the following error when trying to compile mexopencv on OS x 10.9 with MATLAB 2014a:

make MATLABDIR=/Applications/MATLAB_R2014a.app/
/Applications/MATLAB_R2014a.app//bin/mex -cxx -largeArrayDims -Iinclude -I/usr/local/Cellar/opencv/2.4.9/include/opencv -I/usr/local/Cellar/opencv/2.4.9/include  src/+cv/CamShift.cpp -lMxArray -Llib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_calib3d.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_contrib.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_core.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_features2d.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_flann.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_gpu.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_highgui.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_imgproc.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_legacy.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_ml.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_nonfree.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_objdetect.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_ocl.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_photo.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_stitching.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_superres.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_ts.a /usr/local/Cellar/opencv/2.4.9/lib/libopencv_video.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_videostab.dylib  -o +cv/CamShift.mexmaci64
Warning: Legacy MEX infrastructure is provided for compatibility; it will be removed in a future version of MATLAB. For more information, consult the MEX release notes http://www.mathworks.com/help/matlab/release-notes.html.
Undefined symbols for architecture x86_64:
  "cv::merge(std::vector<cv::Mat, std::allocator<cv::Mat> > const&, cv::_OutputArray const&)", referenced from:
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
  "cv::split(cv::Mat const&, std::vector<cv::Mat, std::allocator<cv::Mat> >&)", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status

    mex: link of ' "+cv/CamShift.mexmaci64"' failed.

make: *** [+cv/CamShift.mexmaci64] Error 255

I have the matlab example mex files compiling correctly.  I'm using the the mexopt.sh from this link: http://folk.ntnu.no/seljasen/Arbeider/GSL_and_openMP_mexopts.sh, which specifies gcc and gcc++ (I'm using gcc49).

Any help appreciated.

**Comment by nadacambia** [comment]
Hi, I have the same problem, I tried the different solutions here but nothing. I really need mexopencv to work. I have MATLAB 2014a with OSX 10.10

```
/Applications/MATLAB_R2014a.app/bin/mex -v -c -cxx -largeArrayDims -Iinclude -I/usr/local/include/opencv -I/usr/local/include -I/usr/local/include/opencv2 -I/Applications/MATLAB_R2014a/extern/include -I/usr/local/include/opencv -I/usr/local/include  src/MxArray.cpp -outdir lib
Warning: Legacy MEX infrastructure is provided for compatibility; it will be removed in a future version of MATLAB. For more information, consult the MEX release notes http://www.mathworks.com/help/matlab/release-notes.html.
-> mexopts.sh sourced from directory (DIR = $PREF_DIR)
   FILE = /Users/erik/.matlab/R2014a/mexopts.sh
----------------------------------------------------------------
->    MATLAB                = /Applications/MATLAB_R2014a.app
->    CC                    = xcrun  -sdk macosx10.10  clang
->    CC flags:
         CFLAGS             = -fno-common -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/ -mmacosx-version-min=10.10  -fexceptions -Dchar16_t=uint16_t
         CDEBUGFLAGS        = -g
         COPTIMFLAGS        = -O2 -DNDEBUG
         CLIBS              = /Applications/MATLAB_R2014a.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmat.dylib -lc++
         arguments          = 
->    CXX                   = xcrun  -sdk macosx10.10  clang++
->    CXX flags:
         CXXFLAGS           = -fno-common -fexceptions -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/ -mmacosx-version-min=10.10 -std=gnu++11 -stdlib=libc++
         CXXDEBUGFLAGS      = -g
         CXXOPTIMFLAGS      = -O2 -DNDEBUG
         CXXLIBS            = /Applications/MATLAB_R2014a.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmat.dylib -lc++
         arguments          = 
->    FC                    = gfortran
->    FC flags:
         FFLAGS             = -fexceptions -m64 -fbackslash
         FDEBUGFLAGS        = -g
         FOPTIMFLAGS        = -O
         FLIBS              = /Applications/MATLAB_R2014a.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmat.dylib -L -lgfortran -L -lgfortranbegin
         arguments          = 
->    LD                    = xcrun -sdk macosx10.10 clang
->    Link flags:
         LDFLAGS            = -arch x86_64 -Wl,-syslibroot,/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/ -mmacosx-version-min=10.10 -bundle -Wl,-exported_symbols_list,/Applications/MATLAB_R2014a.app/extern/lib/maci64/mexFunction.map
         LDDEBUGFLAGS       = -g
         LDOPTIMFLAGS       = -O
         LDEXTENSION        = .mexmaci64
         arguments          = 
->    LDCXX                 = 
->    Link flags:
         LDCXXFLAGS         = 
         LDCXXDEBUGFLAGS    = 
         LDCXXOPTIMFLAGS    = 
         LDCXXEXTENSION     = 
         arguments          = 
----------------------------------------------------------------

-> xcrun  -sdk macosx10.10  clang++ -c  -Iinclude -I/usr/local/include/opencv -I/usr/local/include -I/usr/local/include/opencv2 -I/Applications/MATLAB_R2014a/extern/include -I/usr/local/include/opencv -I/usr/local/include -I/Applications/MATLAB_R2014a.app/extern/include -I/Applications/MATLAB_R2014a.app/simulink/include -DMATLAB_MEX_FILE -fno-common -fexceptions -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/ -mmacosx-version-min=10.10 -std=gnu++11 -stdlib=libc++  -O2 -DNDEBUG  "src/MxArray.cpp" -o lib/MxArray.o

ar -cq lib/libMxArray.a lib/*.o
rm -f -f lib/*.o
/Applications/MATLAB_R2014a.app/bin/mex -v -cxx -largeArrayDims -Iinclude -I/usr/local/include/opencv -I/usr/local/include -I/usr/local/include/opencv2 -I/Applications/MATLAB_R2014a/extern/include -I/usr/local/include/opencv -I/usr/local/include  src/+cv/CamShift.cpp -lMxArray -Llib  -output +cv/CamShift.mexmaci64
Warning: Legacy MEX infrastructure is provided for compatibility; it will be removed in a future version of MATLAB. For more information, consult the MEX release notes http://www.mathworks.com/help/matlab/release-notes.html.
-> mexopts.sh sourced from directory (DIR = $PREF_DIR)
   FILE = /Users/erik/.matlab/R2014a/mexopts.sh
----------------------------------------------------------------
->    MATLAB                = /Applications/MATLAB_R2014a.app
->    CC                    = xcrun  -sdk macosx10.10  clang
->    CC flags:
         CFLAGS             = -fno-common -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/ -mmacosx-version-min=10.10  -fexceptions -Dchar16_t=uint16_t
         CDEBUGFLAGS        = -g
         COPTIMFLAGS        = -O2 -DNDEBUG
         CLIBS              = /Applications/MATLAB_R2014a.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmat.dylib -lc++
         arguments          = 
->    CXX                   = xcrun  -sdk macosx10.10  clang++
->    CXX flags:
         CXXFLAGS           = -fno-common -fexceptions -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/ -mmacosx-version-min=10.10 -std=gnu++11 -stdlib=libc++
         CXXDEBUGFLAGS      = -g
         CXXOPTIMFLAGS      = -O2 -DNDEBUG
         CXXLIBS            = /Applications/MATLAB_R2014a.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmat.dylib -lc++
         arguments          = 
->    FC                    = gfortran
->    FC flags:
         FFLAGS             = -fexceptions -m64 -fbackslash
         FDEBUGFLAGS        = -g
         FOPTIMFLAGS        = -O
         FLIBS              = /Applications/MATLAB_R2014a.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmat.dylib -L -lgfortran -L -lgfortranbegin
         arguments          = 
->    LD                    = xcrun -sdk macosx10.10 clang
->    Link flags:
         LDFLAGS            = -arch x86_64 -Wl,-syslibroot,/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/ -mmacosx-version-min=10.10 -bundle -Wl,-exported_symbols_list,/Applications/MATLAB_R2014a.app/extern/lib/maci64/mexFunction.map
         LDDEBUGFLAGS       = -g
         LDOPTIMFLAGS       = -O
         LDEXTENSION        = .mexmaci64
         arguments          =  -lMxArray -Llib
->    LDCXX                 = 
->    Link flags:
         LDCXXFLAGS         = 
         LDCXXDEBUGFLAGS    = 
         LDCXXOPTIMFLAGS    = 
         LDCXXEXTENSION     = 
         arguments          =  -lMxArray -Llib
----------------------------------------------------------------

-> xcrun  -sdk macosx10.10  clang++ -c  -Iinclude -I/usr/local/include/opencv -I/usr/local/include -I/usr/local/include/opencv2 -I/Applications/MATLAB_R2014a/extern/include -I/usr/local/include/opencv -I/usr/local/include -I/Applications/MATLAB_R2014a.app/extern/include -I/Applications/MATLAB_R2014a.app/simulink/include -DMATLAB_MEX_FILE -fno-common -fexceptions -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/ -mmacosx-version-min=10.10 -std=gnu++11 -stdlib=libc++  -O2 -DNDEBUG  "src/+cv/CamShift.cpp"

-> xcrun -sdk macosx10.10 clang -O -arch x86_64 -Wl,-syslibroot,/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/ -mmacosx-version-min=10.10 -bundle -Wl,-exported_symbols_list,/Applications/MATLAB_R2014a.app/extern/lib/maci64/mexFunction.map -o  "+cv/CamShift.mexmaci64"  CamShift.o  -lMxArray -Llib /Applications/MATLAB_R2014a.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014a.app/bin/maci64/libmat.dylib -lc++

Undefined symbols for architecture x86_64:
  "cv::_InputArray::_InputArray(cv::Mat const&)", referenced from:
      _mexFunction in CamShift.o
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
  "cv::_InputArray::_InputArray(double const&)", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
  "cv::_OutputArray::_OutputArray(cv::Mat&)", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
      MxArray::toMatND(int, bool) const in libMxArray.a(MxArray.o)
  "cv::SparseMatConstIterator::SparseMatConstIterator(cv::SparseMat const*)", referenced from:
      MxArray::MxArray(cv::SparseMat const&) in libMxArray.a(MxArray.o)
  "cv::SparseMatConstIterator::operator++()", referenced from:
      MxArray::MxArray(cv::SparseMat const&) in libMxArray.a(MxArray.o)
  "cv::Mat::deallocate()", referenced from:
      _mexFunction in CamShift.o
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
      MxArray::toMatND(int, bool) const in libMxArray.a(MxArray.o)
      std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> > MxArray::toVector<cv::Mat>() const in libMxArray.a(MxArray.o)
      cv::MatExpr::operator cv::Mat() const in libMxArray.a(MxArray.o)
      std::__1::__split_buffer<cv::Mat, std::__1::allocator<cv::Mat>&>::~__split_buffer() in libMxArray.a(MxArray.o)
      ...
  "cv::Mat::setTo(cv::_InputArray const&, cv::_InputArray const&)", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
  "cv::Mat::create(int, int const*, int)", referenced from:
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
  "cv::Mat::copySize(cv::Mat const&)", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
      MxArray::toMatND(int, bool) const in libMxArray.a(MxArray.o)
      std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> > MxArray::toVector<cv::Mat>() const in libMxArray.a(MxArray.o)
      void std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> >::__push_back_slow_path<cv::Mat>(cv::Mat&&) in libMxArray.a(MxArray.o)
      std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> >::__swap_out_circular_buffer(std::__1::__split_buffer<cv::Mat, std::__1::allocator<cv::Mat>&>&) in libMxArray.a(MxArray.o)
  "cv::Mat::Mat(int, int const*, int, void*, unsigned long const*)", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
      MxArray::toMatND(int, bool) const in libMxArray.a(MxArray.o)
  "cv::merge(std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> > const&, cv::_OutputArray const&)", referenced from:
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
  "cv::split(cv::Mat const&, std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> >&)", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
  "cv::Moments::Moments(double, double, double, double, double, double, double, double, double, double)", referenced from:
      MxArray::toMoments(unsigned long) const in libMxArray.a(MxArray.o)
  "cv::compare(cv::_InputArray const&, cv::_InputArray const&, cv::_OutputArray const&, int)", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
  "cv::CamShift(cv::_InputArray const&, cv::Rect_<int>&, cv::TermCriteria)", referenced from:
      _mexFunction in CamShift.o
  "cv::fastFree(void*)", referenced from:
      _mexFunction in CamShift.o
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
      MxArray::toMatND(int, bool) const in libMxArray.a(MxArray.o)
      std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> > MxArray::toVector<cv::Mat>() const in libMxArray.a(MxArray.o)
      cv::MatExpr::operator cv::Mat() const in libMxArray.a(MxArray.o)
      std::__1::__split_buffer<cv::Mat, std::__1::allocator<cv::Mat>&>::~__split_buffer() in libMxArray.a(MxArray.o)
      ...
  "cv::SparseMat::ptr(int, int, bool, unsigned long*)", referenced from:
      MxArray::toSparseMat() const in libMxArray.a(MxArray.o)
  "cv::SparseMat::create(int, int const*, int)", referenced from:
      MxArray::toSparseMat() const in libMxArray.a(MxArray.o)
  "cv::Mat::t() const", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
      MxArray::toMatND(int, bool) const in libMxArray.a(MxArray.o)
  "cv::Mat::convertTo(cv::_OutputArray const&, int, double, double) const", referenced from:
      MxArray::MxArray(cv::Mat const&, mxClassID, bool) in libMxArray.a(MxArray.o)
      MxArray::toMat(int, bool) const in libMxArray.a(MxArray.o)
      MxArray::toMatND(int, bool) const in libMxArray.a(MxArray.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

    mex: link of ' "+cv/CamShift.mexmaci64"' failed.

make: *** [+cv/CamShift.mexmaci64] Error 255
```

**Comment by amroamroamro** [comment]
@nadacambia:
The `mex` calls you've shown do not link against OpenCV libraries, hence the undefined symbol errors (it should include arguments like `/usr/local/lib/libopencv_*.dylib`)...

Maybe it's an issue with `pkg-config`? What do you get when you run the following in a terminal:

```
pkg-config --libs opencv
```
vs.
```
pkg-config --libs-only-L --libs-only-l opencv
```

**Comment by nadacambia** [comment]
That's what I thought, I just don't know how to link it (I tried editing the makefile and adding the opencv directory as in another issue, editing the mexopt.ch and the xml

```
/usr/local/lib/libopencv_calib3d.dylib /usr/local/lib/libopencv_contrib.dylib /usr/local/lib/libopencv_core.dylib /usr/local/lib/libopencv_features2d.dylib /usr/local/lib/libopencv_flann.dylib /usr/local/lib/libopencv_gpu.dylib /usr/local/lib/libopencv_highgui.dylib /usr/local/lib/libopencv_imgproc.dylib /usr/local/lib/libopencv_legacy.dylib /usr/local/lib/libopencv_ml.dylib /usr/local/lib/libopencv_nonfree.dylib /usr/local/lib/libopencv_objdetect.dylib /usr/local/lib/libopencv_ocl.dylib /usr/local/lib/libopencv_photo.dylib /usr/local/lib/libopencv_stitching.dylib /usr/local/lib/libopencv_superres.dylib /usr/local/lib/libopencv_ts.a /usr/local/lib/libopencv_video.dylib /usr/local/lib/libopencv_videostab.dylib 
```

**Comment by amroamroamro** [comment]
@nadacambia: is that the output from the `pkg-config` commands I asked about? if so which one? The Makefile currently uses `--libs-only-L --libs-only-l` to solve an issue with Ubuntu. If that's causing the problem edit it back to use `--libs` instead...

Obviously you can always just manually specify the full OpenCV libraries...

**Comment by kyamagu** [comment]
@nadacambia As @amroamroamro suggests, try editing the lines 16-17 in Makefile to the following.

    C_FLAGS     := -cxx -largeArrayDims -I$(INCLUDEDIR) $(shell pkg-config --cflags opencv)
    LD_FLAGS    := -L$(LIBDIR) $(shell pkg-config --libs opencv)

**Comment by kyamagu** [comment]
@amroamroamro Apparently there is a compatibility issue in `pkg-config`. In Ubuntu's repository package, the default flags `--libs-only-l --libs-only-L` correctly generate:

    $ pkg-config --libs-only-l --libs-only-L opencv
    -lopencv_calib3d -lopencv_contrib -lopencv_core -lopencv_features2d -lopencv_flann -lopencv_gpu -lopencv_highgui -lopencv_imgproc -lopencv_legacy -lopencv_ml -lopencv_objdetect -lopencv_ocl -lopencv_photo -lopencv_stitching -lopencv_superres -lopencv_ts -lopencv_video -lopencv_videostab 

However, `pkg-config` doesn't work with `--libs` flag in Ubuntu (Unknown file extension error), and doesn't apply to other UNIX platforms including a custom build. The [default `opencv.pc`](https://github.com/Itseez/opencv/blob/master/cmake/templates/opencv-XXX.pc.in) simply doesn't generate `-lopencv_*` flags but a list of shared libs. I don't remember if that worked in the past, but I suspect this incompatibility is due to some change in MEX since Matlab R2014a.

For now, I will add an ad-hoc approach to fix this...

**Comment by kyamagu** [comment]
@nadacambia You can try the latest `mexopencv` commit. I have added an ad-hoc solution to fix the linker error.

**Comment by nadacambia** [comment]
@kyamagu 

I edited the lines and it works! Thanks a lot, I have been trying to make it work for weeks!

**Comment by amroamroamro** [comment]
@kyamagu:

Your fix using `sed` can interfere with otherwise valid input. For example:

```sh
$ echo -L/usr/local/lib -lopencv_core | sed -E "s|(/[^/ ]+)+/lib|-l|g" | sed -E "s|\.[^ ]+||g"
```

incorrectly gives: `-L-l -lopencv_core`

---

I think using `sed` or `awk` is unnecessary here, a better solution would be to use Make's own [text transformation functions](http://www.gnu.org/software/make/manual/make.html#Functions). Here is an example I came up with:

### /tmp/Makefile

``` Makefile
CV_LDFLAGS    := $(shell PKG_CONFIG_PATH=/tmp pkg-config --libs opencv)
LIBS_SUFFIXES := %.so %.dylib %.a %.la %.dll.a %.dll
CV_LDFLAGS    := $(filter-out $(LIBS_SUFFIXES),$(CV_LDFLAGS)) \
                 $(patsubst libopencv_%, -lopencv_%, \
                 $(basename $(notdir $(filter $(LIBS_SUFFIXES),$(CV_LDFLAGS)))))

all:
	@echo $(CV_LDFLAGS)
```

### /tmp/opencv.pc

```
# Package Information for pkg-config

prefix=/usr/local
exec_prefix=${prefix}/bin
libdir=${prefix}/lib

Name: OpenCV
Description: Open Source Computer Vision Library
Version: 2.4.10
Libs: -L${libdir} ${libdir}/libopencv_calib3d.so ${libdir}/libopencv_contrib.so ${libdir}/libopencv_core.so ${libdir}/libopencv_features2d.so ${libdir}/libopencv_flann.so ${libdir}/libopencv_gpu.so ${libdir}/libopencv_highgui.so ${libdir}/libopencv_imgproc.so ${libdir}/libopencv_legacy.so ${libdir}/libopencv_ml.so ${libdir}/libopencv_nonfree.so ${libdir}/libopencv_objdetect.so ${libdir}/libopencv_photo.so ${libdir}/libopencv_stitching.so ${libdir}/libopencv_superres.so ${libdir}/libopencv_ts.so ${libdir}/libopencv_video.so ${libdir}/libopencv_videostab.so
Cflags: -I${prefix}/include/opencv -I${prefix}/include
```

### Output
``` sh
$ cd /tmp
$ make
-L/usr/local/lib -lopencv_calib3d -lopencv_contrib -lopencv_core -lopencv_features2d -lopencv_flann -lopencv_gpu -lopencv_highgui -lopencv_imgproc -lopencv_legacy -lopencv_ml -lopencv_nonfree -lopencv_objdetect -lopencv_photo -lopencv_stitching -lopencv_superres -lopencv_ts -lopencv_video -lopencv_videostab
```

**Explanation**:
First I use `filter` and `filter-out` functions to separate the arguments containing shared libraries directly specified as files (as in `file.so`, `file.dylib`, etc...). For those files, I extract the base filename without the extension, and replace `libopencv_*` pattern with the correct `-lopencv_*`. Finally I merge the two parts back together.

If you like this solution, I'll create a pull request with the necessary changes :smiley:

**Comment by kyamagu** [comment]
@amroamroamro You're right, `-L/path/..` can cause a problem. And perhaps using Makefile's own function is more appropriate. I'd appreciate if you can create a patch.

**Comment by amroamroamro** [comment]
@kyamagu:
created PR #133

**Comment by kyamagu** [comment]
@amroamroamro Thanks!

**Comment by kyamagu** [comment]
Hmm, there is another issue that a package-built `opencv.pc` does not produce a `-L` flag, but instead solely relies on the absolute path to individual shared objects. That is, instead of

    -L/path/to/lib -libopencv_calib3d -lopencv_core ...

the `pkg-config --libs opencv` result is

   /path/to/lib/libopencv_calid3d.so /path/to/lib/libopencv_core.so ...

This can cause a linker error if the library path is not in a standard location. Maybe we need to also add `-L` flag in Makefile.

**Comment by kyamagu** [comment]
@amroamroamro Could you check if the following work in your environment?

    CV_LDFLAGS    := $(filter-out $(LIBS_SUFFIXES),$(CV_LDFLAGS)) \
                     $(patsubst libopencv_%, -lopencv_%, \
                     $(basename $(notdir $(filter $(LIBS_SUFFIXES),$(CV_LDFLAGS))))) \
                     -L$(sort $(dir $(filter $(LIBS_SUFFIXES),$(CV_LDFLAGS))))

**Comment by amroamroamro** [comment]
@kyamagu:

I decided to test this as well, so I fired a VM of Ubuntu, and installed the prebuilt [OpenCV package](http://packages.ubuntu.com/trusty/libopencv-dev) (available in *universe* repository):

``` sh
$ lsb_release -a
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.1 LTS
Release:        14.04
Codename:       trusty

$ sudo apt-get install libopencv-dev
```

Here is the output of running `pkg-config` (I'm only showing the `opencv_core` library, the rest of the libraries are similar):

``` sh
$ pkg-config --cflags opencv
-I/usr/include/opencv

$ pkg-config --libs opencv
/usr/lib/x86_64-linux-gnu/libopencv_core.so [...] -lopencv_core [...]

$ pkg-config --libs-only-l opencv
/usr/lib/x86_64-linux-gnu/libopencv_core.so [...]

$ pkg-config --libs-only-L opencv
# (empty: nothing was returned)

$ pkg-config --libs-only-other opencv
-lopencv_core [...]
```

As you pointed out, `pkg-config --libs` does not include a library search path (as in `-L/path/to/libs`). But that's not really a problem here, because by default [`pkg-config`](http://linux.die.net/man/1/pkg-config) strips the standard include and library search paths from the result (this can controlled by setting the `PKG_CONFIG_ALLOW_SYSTEM_CFLAGS` and `PKG_CONFIG_ALLOW_SYSTEM_LIBS ` environment variables)

In fact if you inspect the `opencv.pc` file installed, you'll find the expected `-L${libdir}` part in there, it just got stripped from the pkg-config output because it points to a standard location searched by default:

```
$ cat /usr/lib/x86_64-linux-gnu/pkgconfig/opencv.pc
# Package Information for pkg-config

prefix=/usr
exec_prefix=${prefix}
libdir=${prefix}/lib/x86_64-linux-gnu
includedir_old=${prefix}/include/opencv
includedir_new=${prefix}/include

Name: OpenCV
Description: Open Source Computer Vision Library
Version: 2.4.8
Libs: -L${libdir} ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_calib3d.so -lopencv_calib3d ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_contrib.so -lopencv_contrib ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_core.so -lopencv_core ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_features2d.so -lopencv_features2d ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_flann.so -lopencv_flann ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_gpu.so -lopencv_gpu ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_highgui.so -lopencv_highgui ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_imgproc.so -lopencv_imgproc ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_legacy.so -lopencv_legacy ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_ml.so -lopencv_ml ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_objdetect.so -lopencv_objdetect ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_ocl.so -lopencv_ocl ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_photo.so -lopencv_photo ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_stitching.so -lopencv_stitching ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_superres.so -lopencv_superres ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_ts.so -lopencv_ts ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_video.so -lopencv_video ${exec_prefix}/lib/x86_64-linux-gnu/libopencv_videostab.so -lopencv_videostab
Cflags: -I${includedir_old} -I${includedir_new}
```

so no need to explicitly add the `-L` flag as you did.. It should work whether OpenCV is installed to default location or otherwise.

---

Oviously the main issue is that `pkg-config --libs` above incorrectly returns both the full path to the shared libraries `/path/to/libopencv_*.so`, as well as the `-l` flags. Normally only the `-lopencv_*` part should be returned, so I consider this a bug that needs to be reported to the [Ubuntu package maintainers](https://bugs.launchpad.net/ubuntu/+source/opencv) (there's even [a mention of this bug](https://answers.launchpad.net/ubuntu/+source/opencv/+question/210906) some years ago). Ideally a fix there would solve the issue, making the current ad-hoc solution unneeded!

I believe the problem is mainly in [`OpenCVGenPkgconfig.cmake`](https://github.com/Itseez/opencv/blob/2.4/cmake/OpenCVGenPkgconfig.cmake) which generate the final `opencv.pc` during the build process. I'm not a Debian package expert, but there seems to be a patch in the Ubuntu repository that's supposed to fix the cmake file (see [here](http://bazaar.launchpad.net/~ubuntu-branches/ubuntu/vivid/opencv/vivid/view/head:/debian/patches/pkg-config)). Maybe it's related...

Interestingly, there's also a [recent pull request](https://github.com/Itseez/opencv/pull/3305) in the upstream Itseez/opencv that was merged about 2 months ago (I don't think it made it into Ubuntu repo yet); It refactored the way `opencv.pc` is generated. Perhaps this solves all the above problems (again my CMake knowledge is very limited, so I'm not sure)..

---

Anyways my proposed fix should work for now. I would still check to see whether it works on OSX. I don't have a Mac to verify, so I'm wondering what each package manager would install with respect to the `opencv.pc` file (Fink vs. MacPorts vs. Homebrew if that makes a difference?), and whether the "system-location-stripping" by `pkg-config` applies as well...

**Comment by kyamagu** [comment]
@amroamroamro I just tested this in Mac + Homebrew, and the result is the following:

    $ pkg-config --libs opencv
    /usr/local/Cellar/opencv/2.4.9/lib/libopencv_calib3d.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_contrib.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_core.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_features2d.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_flann.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_gpu.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_highgui.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_imgproc.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_legacy.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_ml.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_nonfree.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_objdetect.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_ocl.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_photo.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_stitching.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_superres.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_ts.a /usr/local/Cellar/opencv/2.4.9/lib/libopencv_video.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_videostab.dylib 

    $ PKG_CONFIG_ALLOW_SYSTEM_LIBS=1 pkg-config --libs opencv
    /usr/local/Cellar/opencv/2.4.9/lib/libopencv_calib3d.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_contrib.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_core.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_features2d.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_flann.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_gpu.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_highgui.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_imgproc.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_legacy.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_ml.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_nonfree.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_objdetect.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_ocl.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_photo.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_stitching.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_superres.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_ts.a /usr/local/Cellar/opencv/2.4.9/lib/libopencv_video.dylib /usr/local/Cellar/opencv/2.4.9/lib/libopencv_videostab.dylib

This unfortunately causes a build error by default because there is no `-L/usr/local/lib`... I also tried to build a source package but the built `opencv.pc` did the same: no library path for some reason. So, at least the current fix still causes an issue in Mac...

As you say, I believe this is a bug in OpenCV CMake. The real fix should be made in OpenCV. For now, let's just add one more nasty fix in Makefile...

**Comment by amroamroamro** [comment]
@kyamagu:
in that case, I'd say go ahead with the fix. Only I would put the  `-L` part before the rest of the `-l` flags, as in:

``` Makefile
# append OpenCV linking flags
LIB_SUFFIX  := %.so %.dylib %.a %.la %.dll.a %.dll
CV_LDFLAGS  := $(shell pkg-config --libs opencv)
CV_LDFLAGS  := $(filter-out $(LIB_SUFFIX),$(CV_LDFLAGS)) \
               -L$(sort $(dir $(filter $(LIBS_SUFFIXES),$(CV_LDFLAGS)))) \
               $(patsubst libopencv_%, -lopencv_%, \
               $(basename $(notdir $(filter $(LIB_SUFFIX),$(CV_LDFLAGS)))))
LD_FLAGS    += $(CV_LDFLAGS)
```

**Comment by amroamroamro** [comment]
@kyamagu
Question: does `mex` (and hence `clang` on a OSX) directly accept a full path to a .dylib file as a linking syntax?

```
mex file.cpp -L/path/to -lfoo
vs.
mex file.cpp /path/to/libfoo.dylib
```

If it does, then there is no need to fix `pkg-config --libs opencv` on Mac, and perhaps we should just remove `%.dylib` from `LIB_SUFFIX` variable in the above snippet, causing it to be used unchanged..

Can you please confirm?

**Comment by kyamagu** [comment]
@amroamroamro Using my environment (OS X 10.10, Homebrew, Matlab R2011a), `mex` directly works with the objects (I needed to remove `.dylib` and `.a`). The question is, if the same applies to Matlab R2014a or above, because I suspect the recent change in `mex` infrastructure is the fundamental cause in not accepting shared objects. I can try to install Matlab R2014a. Just a couple of moments please...

**Comment by amroamroamro** [comment]
@kyamagu 
I see. I guess we should just keep the fix for all platforms then. After all we know that the syntax `-L/path/to -lfoo` works everywhere...

I assume the Makefile works in your environment with the [latest changes](https://github.com/kyamagu/mexopencv/issues/110#issuecomment-70230046) using `addprefix`? If so, we should be good then :smile:

**Comment by kyamagu** [comment]
@amroamroamro Yeah, I can confirm the latest change with `addprefix` solves the problem in Mac. This is pretty annoying... Anyway, great thanks to your contribution. I'll push the change.

**Comment by haipz** [comment]
I read all the comments, and finally I got this below:
<code>
>> mexopencv.make
make MATLABDIR="/Applications/MATLAB_R2014b.app" MEXEXT=mexmaci64  
/Applications/MATLAB_R2014b.app/bin/mex -c -cxx -largeArrayDims -Iinclude -I/usr/local/include/opencv -I/usr/local/include  -v src/MxArray.cpp -outdir lib
警告: 为了实现兼容性而提供了旧的 MEX 基础结构；在以后的 MATLAB 版本中将会删除该基础结构。有关详细信息，请查阅 MEX 发行说明 http://www.mathworks.com/help/matlab/release-notes.html。
/Users/haipz/.matlab/R2014b/mexopts.sh: line 134: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/: is a directory
-> mexopts.sh sourced from directory (DIR = $PREF_DIR)
   FILE = /Users/haipz/.matlab/R2014b/mexopts.sh
----------------------------------------------------------------
->    MATLAB                = /Applications/MATLAB_R2014b.app
->    CC                    = xcrun  -sdk macosx10.10  clang
->    CC flags:
         CFLAGS             = -fno-common -arch x86_64 -isysroot  -mmacosx-version-min=10.10  -fexceptions -Dchar16_t=uint16_t
         CDEBUGFLAGS        = -g
         COPTIMFLAGS        = -O2 -DNDEBUG
         CLIBS              = /Applications/MATLAB_R2014b.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014b.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014b.app/bin/maci64/libmat.dylib -lstdc++
         arguments          = 
->    CXX                   = xcrun  -sdk macosx10.10  clang++
->    CXX flags:
         CXXFLAGS           = -fno-common -fexceptions -arch x86_64 -isysroot  -mmacosx-version-min=10.10 -std=c++11 -stdlib=libc++
         CXXDEBUGFLAGS      = -g
         CXXOPTIMFLAGS      = -O2 -DNDEBUG
         CXXLIBS            = /Applications/MATLAB_R2014b.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014b.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014b.app/bin/maci64/libmat.dylib -lstdc++
         arguments          = 
->    FC                    = gfortran
->    FC flags:
         FFLAGS             = -fexceptions -m64 -fbackslash
         FDEBUGFLAGS        = -g
         FOPTIMFLAGS        = -O
         FLIBS              = /Applications/MATLAB_R2014b.app/bin/maci64/libmx.dylib /Applications/MATLAB_R2014b.app/bin/maci64/libmex.dylib /Applications/MATLAB_R2014b.app/bin/maci64/libmat.dylib -L/usr/local/Cellar/gcc/4.9.2/lib/gcc/x86_64-apple-darwin14.0.0/4.9.2 -lgfortran -L/usr/local/Cellar/gcc/4.9.2/lib/gcc/x86_64-apple-darwin14.0.0/4.9.2 -lgfortranbegin
         arguments          = 
->    LD                    = xcrun -sdk macosx10.10 clang
->    Link flags:
         LDFLAGS            = -arch x86_64 -Wl,-syslibroot, -mmacosx-version-min=10.10 -bundle -Wl,-exported_symbols_list,/Applications/MATLAB_R2014b.app/extern/lib/maci64/mexFunction.map
         LDDEBUGFLAGS       = -g
         LDOPTIMFLAGS       = -O
         LDEXTENSION        = .mexmaci64
         arguments          = 
->    LDCXX                 = 
->    Link flags:
         LDCXXFLAGS         = 
         LDCXXDEBUGFLAGS    = 
         LDCXXOPTIMFLAGS    = 
         LDCXXEXTENSION     = 
         arguments          = 
----------------------------------------------------------------

-> xcrun  -sdk macosx10.10  clang++ -c  -Iinclude -I/usr/local/include/opencv -I/usr/local/include -I/Applications/MATLAB_R2014b.app/extern/include -I/Applications/MATLAB_R2014b.app/simulink/include -DMATLAB_MEX_FILE -fno-common -fexceptions -arch x86_64 -isysroot  -mmacosx-version-min=10.10 -std=c++11 -stdlib=libc++  -O2 -DNDEBUG  "src/MxArray.cpp" -o lib/MxArray.o

clang: warning: no such sysroot directory: '-mmacosx-version-min=10.10'
In file included from src/MxArray.cpp:6:
In file included from include/MxArray.hpp:9:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/functional:477:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:604:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/iterator:335:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/iosfwd:90:10: fatal error: 
      'wchar.h' file not found
#include <wchar.h>  // for mbstate_t
         ^
1 error generated.

    mex: compile of ' "src/MxArray.cpp"' failed.

make: *** [lib/libMxArray.a] Error 255
</code>
I don't know how to do next step.

**Comment by haipz** [comment]
I run the code below:
```
>> mexopencv.make('extra','-I /usr/local/Cellar/opencv/2.4.11/include -L /usr/local/Cellar/opencv/2.4.11/lib/')
```
and I got the result:
```
make MATLABDIR="/Applications/MATLAB_R2014b.app" MEXEXT=mexmaci64  -I /usr/local/Cellar/opencv/2.4.11/include -L /usr/local/Cellar/opencv/2.4.11/lib/
make: Nothing to be done for `/usr/local/Cellar/opencv/2.4.11/lib/'.
```
Then how to solve this problem please? T . T

**Comment by kyamagu** [comment]
@haipz What do you get if you type the following in UNIX shell (not in matlab)

make MATLABDIR="/Applications/MATLAB_R2014b.app" MEXEXT=mexmaci64  -I/usr/local/Cellar/opencv/2.4.11/include -L/usr/local/Cellar/opencv/2.4.11/lib

**Comment by haipz** [comment]
I run the code in item2, and got this:
```
/Applications/Xcode.app/Contents/Developer/usr/bin/make: invalid option -- /
/Applications/Xcode.app/Contents/Developer/usr/bin/make: invalid option -- u
/Applications/Xcode.app/Contents/Developer/usr/bin/make: invalid option -- /
Usage: make [options] [target] ...
Options:
  -b, -m                      Ignored for compatibility.
  -B, --always-make           Unconditionally make all targets.
  -C DIRECTORY, --directory=DIRECTORY
                              Change to DIRECTORY before doing anything.
  -d                          Print lots of debugging information.
  --debug[=FLAGS]             Print various types of debugging information.
  -e, --environment-overrides
                              Environment variables override makefiles.
  -f FILE, --file=FILE, --makefile=FILE
                              Read FILE as a makefile.
  -h, --help                  Print this message and exit.
  -i, --ignore-errors         Ignore errors from commands.
  -I DIRECTORY, --include-dir=DIRECTORY
                              Search DIRECTORY for included makefiles.
  -j [N], --jobs[=N]          Allow N jobs at once; infinite jobs with no arg.
  -k, --keep-going            Keep going when some targets can't be made.
  -l [N], --load-average[=N], --max-load[=N]
                              Don't start multiple jobs unless load is below N.
  -L, --check-symlink-times   Use the latest mtime between symlinks and target.
  -n, --just-print, --dry-run, --recon
                              Don't actually run any commands; just print them.
  -o FILE, --old-file=FILE, --assume-old=FILE
                              Consider FILE to be very old and don't remake it.
  -p, --print-data-base       Print make's internal database.
  -q, --question              Run no commands; exit status says if up to date.
  -r, --no-builtin-rules      Disable the built-in implicit rules.
  -R, --no-builtin-variables  Disable the built-in variable settings.
  -s, --silent, --quiet       Don't echo commands.
  -S, --no-keep-going, --stop
                              Turns off -k.
  -t, --touch                 Touch targets instead of remaking them.
  -v, --version               Print the version number of make and exit.
  -w, --print-directory       Print the current directory.
  --no-print-directory        Turn off -w, even if it was turned on implicitly.
  -W FILE, --what-if=FILE, --new-file=FILE, --assume-new=FILE
                              Consider FILE to be infinitely new.
  --warn-undefined-variables  Warn when an undefined variable is referenced.
  -N OPTION, --NeXT-option=OPTION
                              Turn on value of NeXT OPTION.

This program built for i386-apple-darwin11.3.0
Report bugs to <bug-make@gnu.org>
```

**Comment by kyamagu** [comment]
Oops, that was not the solution. Basically you cannot have whitespace between `-I`, `-L` options and paths. 

    mexopencv.make('extra','-I/usr/local/Cellar/opencv/2.4.11/include -L/usr/local/Cellar/opencv/2.4.11/lib/')

**Comment by haipz** [comment]
Oh, you are right.
I run this in matlab:
```
mexopencv.make('extra','-I/usr/local/Cellar/opencv/2.4.11/include -L/usr/local/Cellar/opencv/2.4.11/lib/')
```
and got the same error:
```
make MATLABDIR="/Applications/MATLAB_R2014b.app" MEXEXT=mexmaci64  -I/usr/local/Cellar/opencv/2.4.11/include -L/usr/local/Cellar/opencv/2.4.11/lib/
/Applications/Xcode.app/Contents/Developer/usr/bin/make: invalid option -- /
/Applications/Xcode.app/Contents/Developer/usr/bin/make: invalid option -- u
/Applications/Xcode.app/Contents/Developer/usr/bin/make: invalid option -- /
Usage: make [options] [target] ...
Options:
  -b, -m                      Ignored for compatibility.
  -B, --always-make           Unconditionally make all targets.
  -C DIRECTORY, --directory=DIRECTORY
                              Change to DIRECTORY before doing anything.
  -d                          Print lots of debugging information.
  --debug[=FLAGS]             Print various types of debugging information.
  -e, --environment-overrides
                              Environment variables override makefiles.
  -f FILE, --file=FILE, --makefile=FILE
                              Read FILE as a makefile.
  -h, --help                  Print this message and exit.
  -i, --ignore-errors         Ignore errors from commands.
  -I DIRECTORY, --include-dir=DIRECTORY
                              Search DIRECTORY for included makefiles.
  -j [N], --jobs[=N]          Allow N jobs at once; infinite jobs with no arg.
  -k, --keep-going            Keep going when some targets can't be made.
  -l [N], --load-average[=N], --max-load[=N]
                              Don't start multiple jobs unless load is below N.
  -L, --check-symlink-times   Use the latest mtime between symlinks and target.
  -n, --just-print, --dry-run, --recon
                              Don't actually run any commands; just print them.
  -o FILE, --old-file=FILE, --assume-old=FILE
                              Consider FILE to be very old and don't remake it.
  -p, --print-data-base       Print make's internal database.
  -q, --question              Run no commands; exit status says if up to date.
  -r, --no-builtin-rules      Disable the built-in implicit rules.
  -R, --no-builtin-variables  Disable the built-in variable settings.
  -s, --silent, --quiet       Don't echo commands.
  -S, --no-keep-going, --stop
                              Turns off -k.
  -t, --touch                 Touch targets instead of remaking them.
  -v, --version               Print the version number of make and exit.
  -w, --print-directory       Print the current directory.
  --no-print-directory        Turn off -w, even if it was turned on implicitly.
  -W FILE, --what-if=FILE, --new-file=FILE, --assume-new=FILE
                              Consider FILE to be infinitely new.
  --warn-undefined-variables  Warn when an undefined variable is referenced.
  -N OPTION, --NeXT-option=OPTION
                              Turn on value of NeXT OPTION.

This program built for i386-apple-darwin11.3.0
Report bugs to <bug-make@gnu.org>
```

**Comment by kyamagu** [comment]
@haipz Hold on a second. You don't need to specify any extra in OS X. Just make sure your `pkg-config` is working fine.

**Comment by haipz** [comment]
I reinstall the `pkg-config` by the code below:
```
sudo brew uninstall pkg-config
sudo brew install pkg-config
sudo brew link --overwrite pkg-config
```
and check the path of `pkg-config`, the content of file `~/.bash_profile` is:
```

# Setting PATH for Python 3.4
# The orginal version is saved in .bash_profile.pysave
PATH="/Library/Frameworks/Python.framework/Versions/3.4/bin:${PATH}"
export PATH

# Setting PATH for Python 2.7
# The orginal version is saved in .bash_profile.pysave
PATH="/Library/Frameworks/Python.framework/Versions/2.7/bin:${PATH}"
export PATH

export PKG_CONFIG_PATH="/usr/X11/lib/pkgconfig:/usr/local/Cellar/cairo/1.14.2/lib/pkgconfig"
```
Is it any wrong?

**Comment by kyamagu** [comment]
@haipz I don't know your specific setting, but if things are set up correctly, you would see compiler flags by

    pkg-config --cflags --libs opencv

If nothing is showing up, you probably have messed up your environment.

**Comment by haipz** [comment]
Unluckly, I got this:
```
pkg-config --cflags --libs opencv
Package opencv was not found in the pkg-config search path.
Perhaps you should add the directory containing `opencv.pc'
to the PKG_CONFIG_PATH environment variable
No package 'opencv' found
```
:-( I try my limited knowledge to solve this problem, but it doesn't work.

**Comment by kyamagu** [comment]
@haipz do `brew install opencv` to install OpenCV, or add your `opencv.pc` file somewhere to `PKG_CONFIG_PATH` variable.

**Comment by haipz** [comment]
I solve the `pkg-config` problem.
First of all, I reinstall opencv by use:
```
sudo brew uninstall opencv
sudo brew install opencv
```
Second, I found something wrong with `~/.bash_profile`:
```
export PKG_CONFIG_PATH="/usr/X11/lib/pkgconfig":"/usr/local/Cellar/cairo/1.14.2/lib/pkgconfig":"/usr/local/Cellar/opencv/2.4.11/lib/pkgconfig"
```
Finally, I got this in iTerm2:
```
***MacBook-Pro% source ~/.bash_profile
***MacBook-Pro% echo $PKG_CONFIG_PATH
/usr/X11/lib/pkgconfig:/usr/local/Cellar/cairo/1.14.2/lib/pkgconfig:/usr/local/Cellar/opencv/2.4.11/lib/pkgconfig
***MacBook-Pro% pkg-config --cflags --libs opencv
-I/usr/local/Cellar/opencv/2.4.11/include/opencv -I/usr/local/Cellar/opencv/2.4.11/include -L/usr/local/Cellar/opencv/2.4.11/lib -lopencv_calib3d -lopencv_contrib -lopencv_core -lopencv_features2d -lopencv_flann -lopencv_gpu -lopencv_highgui -lopencv_imgproc -lopencv_legacy -lopencv_ml -lopencv_nonfree -lopencv_objdetect -lopencv_ocl -lopencv_photo -lopencv_stitching -lopencv_superres -lopencv_ts -lopencv_video -lopencv_videostab
```
But it is still wrong with `mexopencv.make('extra','-I/usr/local/Cellar/opencv/2.4.11/include -L/usr/local/Cellar/opencv/2.4.11/lib/')` in matlab. I guess there must something wrong in XCode or cmake.

**Comment by kyamagu** [comment]
@haipz No, you don't need any extra.

    mexopencv.make()

**Comment by haipz** [comment]
I try to make with `mexopencv.make()`, but I got this:
```
>> mexopencv.make()
make MATLABDIR="/Applications/MATLAB_R2014b.app" MEXEXT=mexmaci64  
Package opencv was not found in the pkg-config search path.
Perhaps you should add the directory containing `opencv.pc'
to the PKG_CONFIG_PATH environment variable
No package 'opencv' found
Package opencv was not found in the pkg-config search path.
Perhaps you should add the directory containing `opencv.pc'
to the PKG_CONFIG_PATH environment variable
No package 'opencv' found
/Applications/MATLAB_R2014b.app/bin/mex -c -cxx -largeArrayDims -Iinclude  src/MxArray.cpp -outdir lib
警告: 为了实现兼容性而提供了旧的 MEX 基础结构；在以后的 MATLAB 版本中将会删除该基础结构。有关详细信息，请查阅 MEX 发行说明 http://www.mathworks.com/help/matlab/release-notes.html。
In file included from src/MxArray.cpp:6:
include/MxArray.hpp:14:10: fatal error: 'opencv2/opencv.hpp' file not found
#include "opencv2/opencv.hpp"
         ^
1 error generated.

    mex: compile of ' "src/MxArray.cpp"' failed.

make: *** [lib/libMxArray.a] Error 255
```
and I really don't know why this happen:
```
Package opencv was not found in the pkg-config search path.
Perhaps you should add the directory containing `opencv.pc'
to the PKG_CONFIG_PATH environment variable
No package 'opencv' found
```

**Comment by kyamagu** [comment]
@haipz `mexopencv` is not compatible with Mathworks CV toolbox by default. To workaround, first find conflicting binaries with `otool -L` command and add them to `DYLD_INSERT_LIBRARIES` variable when launching Matlab (See README). The following might give you a sufficient command to launch matlab.

    echo DYLD_INSERT_LIBRARIES=`pkg-config --libs opencv | grep -E -o "/.*\.(so|dylib)" | sed "s/ /:/g"` matlab

**Comment by haipz** [comment]
Sorry, I don't know how to do exactly. There are logs:
```
echo DYLD_INSERT_LIBRARIES=`pkg-config --libs opencv | grep -E -o "/.*\.(so|dylib)" | sed "s/ /:/g"` matlab
DYLD_INSERT_LIBRARIES= matlab
```
```
echo $DYLD_INSERT_LIBRARIES=`pkg-config --libs opencv | grep -E -o "/.*\.(so|dylib)" | sed "s/ /:/g"` matlab
/usr/local/lib/libtiff.5.dylib= matlab
```

**Comment by kyamagu** [comment]
@haipz Your empty result is simply indicating your `pkg-config` is not correctly returning your own OpenCV path. Check back to your environment and fix `PKG_CONFIG_PATH`. If the path is correctly set up, the following command would show you how to launch matlab.

    echo DYLD_INSERT_LIBRARIES=`pkg-config --libs opencv | grep -o -E "/[^ ]*\.(so|dylib)" | xargs echo | sed "s/ /:/g"` /Applications/MATLAB_R2014b.app/bin/matlab

For example, the above returns

    DYLD_INSERT_LIBRARIES=/usr/local/Cellar/opencv/2.4.9/lib/libopencv_calib3d.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_contrib.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_core.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_features2d.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_flann.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_gpu.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_highgui.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_imgproc.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_legacy.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_ml.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_nonfree.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_objdetect.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_ocl.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_photo.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_stitching.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_superres.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_video.dylib:/usr/local/Cellar/opencv/2.4.9/lib/libopencv_videostab.dylib /Applications/MATLAB_R2014b.app/bin/matlab

Note that the above will be different depending on the environment (i.e., version).

Alternative is to move `libopencv_*.dylib` in Matlab's internal path somewhere else. In OS X, these conflicting libraries are at `/Applications/MATLAB_R2014b.app/bin/maci64/`.

**Comment by haipz** [comment]
@kyamagu Thank you for answering.
I check my `pkg-config` by using:
```
pkg-config --libs opencv
```
and get the correct result:
```
-L/usr/local/Cellar/opencv/2.4.11/lib -lopencv_calib3d -lopencv_contrib -lopencv_core -lopencv_features2d -lopencv_flann -lopencv_gpu -lopencv_highgui -lopencv_imgproc -lopencv_legacy -lopencv_ml -lopencv_nonfree -lopencv_objdetect -lopencv_ocl -lopencv_photo -lopencv_stitching -lopencv_superres -lopencv_ts -lopencv_video -lopencv_videostab
```
this seems right. After that I try the code you gave me, and the results is:
```
NO WORK: echo DYLD_INSERT_LIBRARIES=`pkg-config --libs opencv | grep -o -E "/[^ ]*\.(so|dylib)" | xargs echo | sed "s/ /:/g"` /Applications/MATLAB_R2014b.app/bin/matlab

WORK: echo DYLD_INSERT_LIBRARIES=`pkg-config --libs opencv` /Applications/MATLAB_R2014b.app/bin/matlab

NO WROK: echo DYLD_INSERT_LIBRARIES=`grep -o -E "/[^ ]*\.(so|dylib)" | xargs echo | sed "s/ /:/g"` /Applications/MATLAB_R2014b.app/bin/matlab

NO WROK: echo DYLD_INSERT_LIBRARIES=`grep -o -E "/[^ ]*\.(so|dylib)"` /Applications/MATLAB_R2014b.app/bin/matlab

NO WROK: echo DYLD_INSERT_LIBRARIES=`xargs echo` /Applications/MATLAB_R2014b.app/bin/matlab

NO WROK: echo DYLD_INSERT_LIBRARIES=`sed "s/ /:/g"` /Applications/MATLAB_R2014b.app/bin/matlab
```
The only work one respond:
```
DYLD_INSERT_LIBRARIES=-L/usr/local/Cellar/opencv/2.4.11/lib -lopencv_calib3d -lopencv_contrib -lopencv_core -lopencv_features2d -lopencv_flann -lopencv_gpu -lopencv_highgui -lopencv_imgproc -lopencv_legacy -lopencv_ml -lopencv_nonfree -lopencv_objdetect -lopencv_ocl -lopencv_photo -lopencv_stitching -lopencv_superres -lopencv_ts -lopencv_video -lopencv_videostab /Applications/MATLAB_R2014b.app/bin/matlab
```

**Comment by kyamagu** [comment]
@haipz Okay, then you can construct a launcher command by manually specifying all the dependent dylibs, like

    DYLD_INSERT_LIBRARIES=/usr/local/Cellar/opencv/2.4.11/lib/libopencv_calib3d.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_contrib.dylib:... /Applications/MATLAB_R2014b.app/bin/matlab

Note that you must list all the dependencies above (`...`).

**Comment by haipz** [comment]
@kyamagu I run:
```
DYLD_INSERT_LIBRARIES=/usr/local/Cellar/opencv/2.4.11/lib/libopencv_calib3d.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_contrib.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_core.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_features2d.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_flann.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_gpu.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_highgui.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_imgproc.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_legacy.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_ml.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_nonfree.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_objdetect.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_ocl.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_photo.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_stitching.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_superres.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_video.dylib:/usr/local/Cellar/opencv/2.4.11/lib/libopencv_videostab.dylib /Applications/MATLAB_R2014b.app/bin/matlab
```
and here returns a new error:
```
dyld: Library not loaded: /usr/local/lib/libImath-2_1.11.dylib
  Referenced from: /usr/local/Cellar/opencv/2.4.11/lib/libopencv_highgui.dylib
  Reason: image not found
zsh: trace trap  DYLD_INSERT_LIBRARIES= /Applications/MATLAB_R2014b.app/bin/matlab
```

**Comment by kyamagu** [comment]
@haipz I suspect [this issue](https://github.com/Homebrew/homebrew-science/issues/2209).

**Comment by haipz** [comment]
@kyamagu Before I submit my comment, I have read this issue.
```
brew switch ilmbase 2.1.0
Error: ilmbase does not have a version "2.1.0" in the Cellar.
Versions available: 2.2.0
```
I don't have version 2.1.0 in the Cellar. I try to install homebrew-versions to find the older version, but homebrew-versions isn't include `ilmbase 2.1.0`.

**Comment by kyamagu** [comment]
@haipz Great :)
